spring-integration-redis/src/main/java/org/springframework/integration/redis/inbound/ReactiveRedisStreamMessageProducer.java

```diff
@@ -0,0 +1,228 @@
+	 * @param streamReceiverOptions the desired receiver options
+	 * */
+	public void setStreamReceiverOptions(@Nullable StreamReceiver.StreamReceiverOptions<String, ?> streamReceiverOptions) {
+		this.streamReceiverOptions = streamReceiverOptions;
+	}
+
+	@Override
+	public String getComponentType() {
+		return "redis:stream-inbound-channel-adapter";
+	}
+
+	@Override
+	protected void onInit() {
+		super.onInit();
+		this.streamReceiver = StreamReceiver.create(this.reactiveConnectionFactory, this.streamReceiverOptions);
+		if (StringUtils.hasText(this.consumerName) && StringUtils.isEmpty(this.consumerGroup)) {
+			this.consumerGroup = getBeanName();
+		}
+		this.reactiveRedisTemplate = new ReactiveRedisTemplate<>(this.reactiveConnectionFactory,
+				RedisSerializationContext.string());
+		this.reactiveStreamOperations = this.reactiveRedisTemplate.opsForStream();
+	}
+
+	@Override
+	protected void doStart() {
+		super.doStart();
+
+		StreamOffset<String> offset = StreamOffset.create(this.streamKey, this.readOffset);
+
+		Flux<Message<?>> messageFlux;
+
+		if (StringUtils.isEmpty(this.consumerName)) {
+			messageFlux = this.streamReceiver
+					.receive(offset)
+					.map(event -> getMessageBuilderFactory().withPayload(this.extractPayload ? event.getValue() : event)
+							.setHeader(RedisHeaders.STREAM_KEY, event.getStream())
+							.setHeader(RedisHeaders.STREAM_MESSAGE_ID, event.getId())
+							.build());
+		}
+		else {
+			Assert.notNull(this.consumerGroup, "'consumerGroup' must not be null");
+
+			if (this.createConsumerGroup) {
+				Assert.hasText(this.consumerName, "'consumerName' must be set");
+				//We check that the group does not already exist before creating it
+				this.reactiveRedisTemplate.hasKey(this.streamKey)
+						.subscribe(keyExists -> {
+							if (!keyExists) {
+								this.reactiveStreamOperations.createGroup(this.streamKey, this.consumerGroup)
+										.subscribe();
+							}
+							else {
+								this.reactiveStreamOperations.groups(this.streamKey)
+										.map(StreamInfo.XInfoGroup::groupName)
+										.filter(groupName -> groupName.equals(this.consumerGroup))
+										.switchIfEmpty(this.reactiveStreamOperations.createGroup(this.streamKey,
+												this.consumerGroup))
+										.subscribe();
+							}
+						});
+			}
+
+			Consumer consumer = Consumer.from(this.consumerGroup, this.consumerName);
+
+			messageFlux = this.streamReceiver
+					.receive(consumer, offset)
+					.map(event -> {
+						Message<?> message = getMessageBuilderFactory().withPayload(this.extractPayload ? event.getValue() : event)
+								.setHeader(RedisHeaders.STREAM_KEY, event.getStream())
+								.setHeader(RedisHeaders.STREAM_MESSAGE_ID, event.getId())
+								.setHeader(RedisHeaders.CONSUMER_GROUP, this.consumerGroup)
+								.setHeader(RedisHeaders.CONSUMER, this.consumerName)
+								.build();
+						if (this.autoAck) {
```

Comment:
That's something wrong: why do we need to add a manual ack callback header, if we do that when `autoAck`?
Maybe you mean exactly opposite?
Why don't you call `receiveAutoAck()` anymore?

Reply:
How to handle the case where `autoAck` is `false` since `receiveAutoAck()` will always acknowledge?

Reply:
Right, we do `receiveAutoAck()` when `autoAck` is `true` and we add `IntegrationMessageHeaderAccessor.ACKNOWLEDGMENT_CALLBACK` when it is `false`.
And yes: when it is `false`, we perform `receive()`.

When `autoAck == true`, we call `receiveAutoAck()` and don't add header.
We do exactly opposite, when `autoAck == false`.

Reply:
Right, a question. Why adding `IntegrationMessageHeaderAccessor.ACKNOWLEDGMENT_CALLBACK` header when `autoAck==false` since we do not want to acknowledge.

Reply:
Huh? We don't want to ack automatically. With that header we just defer an ack to end-user decision. 
This is typical messaging pattern: when we auto ack, we leave the durability for a message up to end-user, since message is going to be removed from the broker.
If we don't ack and defer to end-user, we still keep a message on the broker and it is going to be available for re-fetch after crash. This way durability is on broker and it is up to end-user to decide when to ack that processed message.

See here more explanation: https://stackoverflow.com/questions/11673361/jms-auto-acknowledge-when-is-it-acknowledged